File: scripts/reference-solutions/maximum-profit-in-job-scheduling.py
import sys
from bisect import bisect_right


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    jobs = []
    idx = 1
    for _ in range(n):
        start = int(data[idx])
        end = int(data[idx + 1])
        profit = int(data[idx + 2])
        idx += 3
        jobs.append((end, start, profit))
    jobs.sort()  # sort by end time

    ends = []
    best = []  # best[i] = max profit achievable using the first i + 1 jobs by end time
    running = 0
    for end, start, profit in jobs:
        pos = bisect_right(ends, start) - 1
        prev = best[pos] if pos >= 0 else 0
        candidate = prev + profit
        if candidate > running:
            running = candidate
        ends.append(end)
        best.append(running)

    print(running)


main()

File: scripts/reference-solutions/maximum-profit-in-job-scheduling.cpp
#include <algorithm>
#include <array>
#include <iostream>
#include <vector>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int n;
  cin >> n;
  vector<array<long long, 3>> jobs(n);  // {end, start, profit}
  for (int i = 0; i < n; ++i) {
    long long start, end, profit;
    cin >> start >> end >> profit;
    jobs[i] = {end, start, profit};
  }
  sort(jobs.begin(), jobs.end());  // sort by end time

  vector<long long> ends(n);
  vector<long long> best(n);  // best[i] = max profit using the first i + 1 jobs
  long long running = 0;
  for (int i = 0; i < n; ++i) {
    long long end = jobs[i][0];
    long long start = jobs[i][1];
    long long profit = jobs[i][2];
    // Largest index whose end time <= start (jobs that finish before this one starts).
    int lo = 0, hi = i - 1, pos = -1;
    while (lo <= hi) {
      int mid = lo + (hi - lo) / 2;
      if (ends[mid] <= start) {
        pos = mid;
        lo = mid + 1;
      } else {
        hi = mid - 1;
      }
    }
    long long prev = pos >= 0 ? best[pos] : 0;
    long long candidate = prev + profit;
    running = max(running, candidate);
    ends[i] = end;
    best[i] = running;
  }

  cout << running << '\n';
  return 0;
}
